### PR TITLE
Add storage engine metrics

### LINKED ISSUE #885: Convert InfluxDB storage engine metrics to prometheus metrics
As we bring the old storage engine code in from InfluxDB we will want to convert all the the metrics that used to report to the `_internal` database into prometheus metrics, to complement our existing set of metrics.

### PR DESCRIPTION
**Pull request by e-dard**
- [ ] CHANGELOG.md updated with a link to the PR (not the Issue)
  - [ ] Rebased/mergeable
  - [ ] Tests pass
  - [ ] swagger.json updated (if modified Go structs or API)
  - [ ] Sign [CLA](https://influxdata.com/community/cla/) (if not already signed)

Closes #885

_Briefly describe your proposed changes:_

_What was the problem?_
We didn't have any prometheus metrics for the storage engine, series file or index.

_What was the solution?_

This PR adds prometheus metrics for the following areas:

**Compactions**
  - Number of compactions by level;
  - Compactions currently actively;
  - Histogram of compaction durations by level;
  - Compactions currently queued up.

**TSM Files**
  - Number of bytes TSM files using on disk;
  - Number of TSM files.

**TSM Cache**
  - Heap size;
  - On disk size (snapshot file);
  - Number of active snapshote (can be greater than 1 for big caches);
  - Time since we last snapshotted cache;
  - Total number of bytes snapshotted;
  - Total number of bytes written to the cache;
  - Number of write calls on the cache.

**WAL**
  - Number of bytes in old segments (full WAL files);
  - Number of bytes in the current active segment file;
  - Number of segment files;
  - Total number of write calls to the WAL.

**Series File**
  All metrics are by partition.
  - Created series;
  - Total series;
  - Disk size of series files;
  - Number of segment files in series file;
  - Active series index compactions; 
  - Histogram of series index compaction times;
  - Total number of series index compactions.
  - Load factor on series index hashmap;
  - Size in bytes of series index hashmap; 
  - Histogram of GET durations for series index hashmap (sampled 1 in 10 calls);
  - Last GET duration for series index hashmap (sampled 1 in 10 calls);
  - Histogram of PUT durations for series index hashmap (sampled 1 in 10 calls);
  - Last PUT duration for series index hashmap (sampled 1 in 10 calls);
  - Last hashmap grow duration for series index hashmap;
  - Total number of GETs from the series index hashmap (broken out by hits and misses);
  - Total number of PUTs to the series index hashmap (broken out by hits and misses).

**TSI Tag Cache**
  - Number of items (bitsets) in the cache;
  - Number of GETs to the cache (broken out by hits and misses);
  - Number of PUTs to the cache (broken out by hits and misses);
  - Number of DELETEs on the cache;
  - Number of EVICTIONs on the cache (it's an LRU cache).

**TSI Index**
All metrics are by partition.
  - Created series;
  - Histogram of single series creation time (nanoseconds);
  - Series removed;
  - Total series;
  - Total number of files (broken out by index files and log files);
  - Disk size in bytes;
  - Number of active TSI compactions (broken out by level);
  - Histogram of compaction duration (broken out by level);
  - Total number of compactions (broken out by level).

### CODE REVIEW

File: pkg/rhh/metrics.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+			Name:      "get_duration_last_ns",
+			Help:      "Last retrieval duration in nanoseconds (sampled every 10% of retrievals)",
+		}, names),
+		InsertDuration: prometheus.NewHistogramVec(prometheus.HistogramOpts{
+			Namespace: namespace,
+			Subsystem: subsystem,
+			Name:      "put_duration_ns",
+			Help:      "Times taken to insert elements in nanoseconds (sampled every 10% of insertions).",
+			// 15 buckets spaced exponentially between 100 and ~30,000.
+			Buckets: prometheus.ExponentialBuckets(100., 1.5, 15),
```
Comment (jacobmarble) [inline pkg/rhh/metrics.go]: This seems arbitrary. It's a good idea to put name buckets objects and put them in a common location, even if it's just the top of this file.

No need to get that in on this PR, though.

File: tsdb/series_file.go [hunk 2]
```diff
@@ -43,7 +53,9 @@ type SeriesFile struct {
 // NewSeriesFile returns a new instance of SeriesFile.
 func NewSeriesFile(path string) *SeriesFile {
 	return &SeriesFile{
-		path:   path,
+		path: path,
+		// partitionMetrics: newSeriesFileMetrics(nil),
```
Comment (jacobmarble) [inline tsdb/series_file.go]: Delete these comments?

File: tsdb/series_partition.go [hunk 3]
```diff
@@ -570,10 +701,10 @@ func (c *SeriesPartitionCompactor) Compact(p *SeriesPartition) error {
 		}
 		return nil
 	}(); err != nil {
-		return err
+		return 0, err
 	}
 
-	return nil
+	return duration, nil
```
Comment (jacobmarble) [inline tsdb/series_partition.go]: Love this.

File: tsdb/tsi1/cache.go [hunk 4]
```diff
@@ -193,3 +212,58 @@ type seriesIDCacheElement struct {
...
+	for k, v := range t.labels {
+		l[k] = v
+	}
+	return l
+}
+
+func (t *cacheTracker) SetSize(sz uint64) {
+	labels := t.Labels()
+	t.metrics.Size.With(labels).Set(float64(sz))
+}
```
Comment (zeebo) [inline tsdb/tsi1/cache.go]: Copies again! I ended up writing a cache for the labels with an api like

```go
labels := cache.Get("status", status, "foo", "bar")
```

meaning labels with `"status"` set to `status`, and `"foo"` set to `"bar"`. In the hit case it does no allocations and no shared memory writes.

```
BenchmarkCache/None-8  500000000  3.42 ns/op  0 B/op    0 allocs/op
BenchmarkCache/One-8   100000000  18.2 ns/op  0 B/op    0 allocs/op
BenchmarkCache/Two-8   50000000   23.7 ns/op  0 B/op    0 allocs/op
BenchmarkCopy/None-8   10000000   188 ns/op   336 B/op  2 allocs/op
BenchmarkCopy/One-8    10000000   210 ns/op   336 B/op  2 allocs/op
BenchmarkCopy/Two-8    10000000   229 ns/op   336 B/op  2 allocs/op
```

File: tsdb/tsm1/cache.go [hunk 5]
```diff
@@ -178,12 +160,7 @@ type storer interface {
 
 // Cache maintains an in-memory store of Values for a set of keys.
 type Cache struct {
-	// Due to a bug in atomic  size needs to be the first word in the struct, as
-	// that's the only place where you're guaranteed to be 64-bit aligned on a
-	// 32 bit system. See: https://golang.org/pkg/sync/atomic/#pkg-note-BUG
-	size         uint64
-	snapshotSize uint64
-
+	_       uint64 // Padding for 32 bit struct alignment
```
Comment (zeebo) [inline tsdb/tsm1/cache.go]: I'm confused. Why is this necessary? The cache is guaranteed to be 64 bit aligned already, so why add another 64 bit entry at the start? What field is getting atomic'd?

File: tsdb/tsm1/cache.go [hunk 6]
```diff
@@ -777,23 +720,167 @@ func (c *Cache) LastWriteTime() time.Time {
...
+// As well as being responsible for providing atomic reads and writes to the
+// statistics, cacheTracker also mirrors any changes to the external prometheus
+// metrics, which the Engine exposes.
+//
+// *NOTE* - cacheTracker fields should not be directory modified. Doing so
+// could result in the Engine exposing inaccurate metrics.
+type cacheTracker struct {
+	metrics         *cacheMetrics
+	labels          prometheus.Labels
+	snapshotsActive uint64
```
Comment (zeebo) [inline tsdb/tsm1/cache.go]: Speaking of 64 bit alignment, a pointer on 386 is 32 bits, and labels is some map type, so I'm not sure but probably also just a pointer? That would make this ok, but it seems safer to move the uint64s to the top.

File: tsdb/tsm1/engine.go [hunk 7]
```diff
@@ -1106,6 +1016,163 @@ func (e *Engine) CreateSeriesListIfNotExists(collection *tsdb.SeriesCollection)
...
+
+// compactionTracker tracks compactions and snapshots within the Engine.
+//
+// As well as being responsible for providing atomic reads and writes to the
+// statistics tracking the various compaction operations, compactionTracker also
+// mirrors any writes to the prometheus block metrics, which the Engine exposes.
+//
+// *NOTE* - compactionTracker fields should not be directory modified. Doing so
+// could result in the Engine exposing inaccurate metrics.
+type compactionTracker struct {
```
Comment (zeebo) [inline tsdb/tsm1/engine.go]: Same comment w.r.t. 32 bit atomics. Seems safest to put all the uint64s at the top.
